🗒 동기화메모장 [STEP 2] Ryan

ryan-son opened the pull request:
@AppleCEO 

안녕하세요, 도미닉! 꼬박 한 주 반만에 찾아 뵙습니다. 제가 느끼기에 Core Data는 지금까지 학습했던 내용과는 맥락이 다르다고 느껴 시간을 조금 소비했습니다. 이에 더불어 기능이 작동하는지 보다 도미닉과 같은 프로그래머가 합리적이라고 여길 수 있는 구조를 구성하려 노력했습니다. 적지 않은 변경사항이 생긴만큼 리뷰하시기에 앞서 코드를 읽으시는데 도움이 되는 내용을 말씀드리겠습니다.

## 주요 타입의 역할과 요청

사용자의 요청에 따른 기능을 제공하기 위해 구성한 구조는 아래와 같습니다. 
![image](https://user-images.githubusercontent.com/69730931/122114478-484a2b00-ce5e-11eb-83d8-ee0dd5991532.png)


기능 구현을 위해 관여하는 주요한 타입은 총 8 개로, 화면에 보여질 것을 담당하는 클래스 타입 4 개, 코어 데이터와 이를 관리하는 구조체 타입 2 개, 타입 간 소통 창구를 마련해주는 딜리게이트 프로토콜 3 개 입니다. 자세한 내용은 도식에 표현하였으니 참고해주세요.

## 작동 영상
### 스와이프 액션을 통한 공유, 상세 화면에서 더보기 버튼을 이용한 삭제
![ezgif com-gif-maker (1)](https://user-images.githubusercontent.com/69730931/122188829-54200680-cecb-11eb-816f-9c4f5bcc5cb2.gif)

### 작성 시간 순 노트 순서 정렬 및 변경사항 셀에 실시간으로 반영
![ezgif com-gif-maker](https://user-images.githubusercontent.com/69730931/122188870-5eda9b80-cecb-11eb-9b0d-f4d9cb3d28c5.gif)

## 구현 간 고민한 점과 적용한 점 (시간이 충분하지 않으시다면 스킵하셔도 됩니다.)
### 데이터 소스 구성과 개체 특정 방식
1. Step 2에 진입하며 Data라고 이름 붙여진 Core data stack과 Diffable data source가 어떤 타입에 위치하여야 할지에 대해 고민하였습니다. 크게 두 가지 옵션을 고려하였습니다.
- data라고 이름 붙여진 대상들을 하나의 타입에 위치시키고 뷰 컨트롤러에는 보여주는 것만을 담당하는 요소들을 모아 구조화하는 방법
- Core data stack과 diffable data source를 각자 다른 타입에 위치시키는 방법

최초에는 전자의 구조가 더 설득력 있다고 판단하여 구현하였지만, 오히려 diffable data source의 존재로 인해 해당 타입이 뷰와 더 강하게 결합되는 결과를 낳았습니다. Diffable data source는 특정 시점의 데이터를 저장하는 스냅샷을 가지고 있다는 측면에서 데이터 제공자라고 생각하였지만, `init(collectionView:cellProvider:)` 이니셜라이저를 통해 컬렉션 뷰의 셀 모양을 제공하는 뷰의 역할을 담당하고 있었기 때문이었습니다. 부가적으로는 diffable data source과 core data stack을 함께 가진 타입은 테스트하기 굉장히 까다로워진다는 문제점이 있었습니다. 단지 CRUD 기능만을 검증하기를 원하여도 diffable data source를 위해 collectionView를 구성하여 입력해주어야 하기 때문에 테스트 간 뷰 컨트롤러의 인스턴스가 필요하게 되어 테스트가 컬렉션 뷰를 제공하는 뷰 컨트롤러에 의존하게 되었습니다.

이러한 시행착오를 겪고 저는 후자의 구조로 구조 개선을 실시하였습니다. 문제점을 파악하니 Core data stack이 제공하여야 하는 기능은 순수하게 본인의 context에 접근하여 수행하는 CRUD 기능이라는 것을 깨달았습니다. 구조를 개선한 후 core data stack은 fetched results controller 이외에는 추가로 기능 구현을 하지 않고 최대한 원형을 유지하여 Core data stack이라는 도구 역할에 충실하게끔 구성하였습니다. 이후 core data stack의 context에 접근하여 앱에서 발생하는 데이터를 처리하여 CRUD를 수행하는 `NoteManager` 타입을 작성하였습니다. `NoteManager`는 뷰와 결합된 요소가 존재하지 않기 때문에 손쉽게 unit test가 가능하였습니다.

2. 데이터를 불러와 사용하는 방식에 대해 고민하였습니다.

본 프로젝트는 리스트 뷰 형식을 사용하고 있기에 특정 데이터를 `indexPath`를 통해 파악할 수 있습니다. 하지만 저는 수정 시 데이터들이 실시간으로 시간 순으로 정렬되기 때문에 `indexPath`를 이용한 데이터 탐색 방식은 불안정하다고 느꼈고, 결국 모델 타입의 특성을 이용하기로 하였습니다. 앱에서 발생하는 데이터 모델은 `NSManagedObject` 타입이며 `Identifiable` 프로토콜을 채택하고 있습니다. 또한 `NSManagedObjectID` 타입인 `objectID` 프로퍼티를 가지고 있기 때문에 설령 데이터가 수정 되더라도 항상 고유한 개체로 인식할 수 있습니다. 이러한 특징으로 인해 최초 데이터를 인식할 때를 제외하고는 `indexPath`를 통해 데이터를 탐색하게 되는 경우를 방지할 수 있습니다. 또 한 가지 재미있는 점은 고유한 식별자라는 특성으로 인해 특정 데이터는 생성된 순간부터 삭제되는 순간까지 동일한 해시값을 가지기 때문에 diffable data source를 통해서도 특정 데이터를 탐색할 수 있다는 것입니다. 이는 core data stack과 함께 작업하는 `NoteManager`에서는 `FetchedResultsController`의  `managedObjectContext`를 통해, collection view와 함께 작업하는 `UICollectionViewDiffableDataSource`는 적용된 `snapshot`을 통해 `indexPath` 및 `objectID (또는 데이터 자체(해시값))`을 통해 양 방향으로 데이터를 탐색하는 방법을 지원함을 의미합니다. 바꾸어 표현하면, `NoteManager`도, 리스트 뷰를 담당하는 `NoteListViewController`에서도 `indexPath` 및 `각 데이터의 고유값`을 통해 데이터를 탐색할 수 있다는 것입니다. 저는 작업을 구성할 때 context의 저장과 snapshot의 적용 시점을 동일하게 구성하여 항상 동일한 데이터를 가지게끔 구성하였고, 데이터의 탐색도 core data의 역할인 Read에 해당한다고 판단하여 `NoteManager`에 해당 역할을 할당하고자 하였으나, indexPath를 이용하여 데이터를 탐색하는 fetchedResultsController의 `object(at:)` 메서드가 유효하지 않은 indexPath를 입력할 경우 예외가 발생한다는 내용을 파악하고 데이터 탐색 역할을 diffable data source에게 일임하였습니다.
![image](https://user-images.githubusercontent.com/69730931/122128438-a7646b80-ce6f-11eb-9bc7-5ff79a662e20.png)

3. 인스턴스 내부에서 메서드별 인스턴스 프로퍼티 등록 및 참조 시점에 따른 오류를 최소화하기 위해 인스턴스 프로퍼티의 수를 최소화하고, 필요한 경우 항상 최신화된 값을 제공할 수 있는 get-only 형식의 연산 프로퍼티를 사용하였습니다.

textView가 `textViewDidChange(_:)` 메서드가 호출될 때 딜리게이트를 통해 `ListViewController`에 뷰의 업데이트 및 core data에 저장하는 요청을 송신할 때 관련된 update 메서드는 특정 데이터를 인식할 수 있는 수단인 `indexPath` 혹은 데이터 자체를 가지고 있지 않습니다. 이를 인식하기 위해 `ListViewController`에 `editingNote`라는 저장 프로퍼티를 마련하였고, 선택한 데이터를 인식할 수 있는 이벤트가 발생하면 해당 프로퍼티에 등록하도록 구성하여 textView에 의한 업데이트가 수행될 때 항상 수정될 데이터가 무엇인지 알려주는 역할을 수행하게끔 하였습니다.
사용 중 수정할 데이터를 인식할 수 있는 이벤트는 아래와 같습니다.
- 셀 선택으로 인해 `collectionView(_:didSelectItemAt:)` 메서드가 호출된 경우
- Add 버튼을 통한 데이터 생성 시 (생성 메서드에서 반환되는 데이터를 통해 인식 가능)
- 특정 데이터 삭제 후 선택되는 대상 설정 - 리스트 뷰의 최상단 데이터 ([0, 0] indexPath 또는 dataSource 스냅샷의 첫 번째 데이터)

텍스트뷰를 제공하는 `NoteDetailViewController`는 NoteListViewControllerActionsDelegate를 통해 더보기 버튼에서 삭제와 액티비티 뷰 보여주기 기능을 제공합니다. 하지만 문제는 리스트뷰와 달리 해당 액션이 일어날 데이터의 indexPath 값을 알 수 없다는 점입니다. 이를 보완하기 위해 딜리게이트를 통해 `NoteListViewController`로 부터 선택된 데이터의 정보를 가져오는 방법을 고려할 수 있지만, 디테일 뷰는 `NoteManager`와 `DiffableDataSource` 모두를 인식하지 못하기에 indexPath를 가져오거나 입력 받는 방식이 가장 유효합니다. 하지만 `NoteListViewController`는 현재 수정되고 있는 데이터를 저장해두는 `editingNote`와 현재 수정되고 있는 `indexPath`를 모두 저장해두기에는 각 프로퍼티를 중복해서 관리해야 하기 때문에 코드의 유지보수에 좋지 않은 영향을 미칠 수 있다고 판단하였고, diffable data source를 통해 선택된 데이터의  항상 최신화된 indexPath를 제공해줄 수 있도록 get-only 형식의 연산 프로퍼티로 구현하였습니다. 요약하여 표현하면, 현재 수정되는 대상은 `NoteListViewController`에서만 관리하고, 참조되는 대상의 수를 최소화하여 에러를 방지하고자 하였다는 것입니다.

4. 요구사항을 정형화된 테스트 스펙으로 활용하여 검증할 수 있도록 UI 테스트를 적용하였습니다.

테스트 스펙이 정형화되어 있지 않으면 직접 사용해보며 버그의 유무를 검토하는 방법 이외에는 예상치 못한 결과를 파악할 수 있는 뾰족한 방법이 없었습니다. 또한 대부분의 구현 내용이 UI에 위치해 있어 모델의 로직만을 unit test로 검증하기에는 code coverage가 낮다는 점도 의도치 않은 버그가 발생할 수 있기 때문에 문제가 있어 보였습니다.
![image](https://user-images.githubusercontent.com/69730931/122132971-d205f280-ce76-11eb-94d8-7e8f0a61710a.png)
이를 UI 테스트를 통해 아래와 같이 보완하였습니다.
![image](https://user-images.githubusercontent.com/69730931/122133086-0a0d3580-ce77-11eb-9912-7b800298a151.png)

## 질문
1. Unit test의 경우에는 in-memory 방식의 mock 타입을 작성하여 활용하는 방식으로 테스트 시 앱 본체의 core data와는 별도의 환경을 구성하여 사용하였는데, UI test 시에도 이러한 mock 타입 인스턴스를 사용하여 별도의 저장소에서 테스트할 수 있는 방법이 있을까요? UI 테스트로 인해 생성된 데이터가 실제 앱 실행 시에도 남아있어 문제가 있어 보입니다.
- UI 테스트는 unit test와 달리 블랙박스 테스트라는 [내용](https://mildwhale.github.io/2019-12-26-uitesting-tip-and-tricks/)을 봤는데, 별도의 core data 저장소를 만들어 사용할 수 없다는 의미인가요?

2. 코어 데이터를 이용한 CRUD 구현에 있어 현재 조직한 구조 (타입 간의 연관 관계)에 대해 어떻게 생각하시나요? 어떤 부분을 개선해보면 좋을까요?

3. 텍스트를 수정하면 실시간으로 반영이 되게끔 구성하였는데요, 최초에는 Diffable data source의 스냅샷을 데이터 소스에 적용할 때 애니메이션 효과를 주고 싶었습니다. 하지만 셀의 순서 변경과 같은 애니메이션은 작동해도 데이터가 변경된 것이 셀에 반영되지는 않았는데, 어떤 실수를 하고 있는걸까요?
![ezgif com-video-to-gif (11)](https://user-images.githubusercontent.com/69730931/122194471-710b0880-ced0-11eb-8505-0885054be988.gif)
앱을 완전히 종료 후 재실행하고 첫 번째 노트를 탭하였을 때 화면 (코어데이터에서 정상적으로 데이터를 로드하여 셀에 반영됨)
![image](https://user-images.githubusercontent.com/69730931/122195064-f989a900-ced0-11eb-93c8-70065ff40bb7.png)

이번 리뷰도 미리 감사드립니다! 😄 오늘도 즐거운 하루 보내세요~ 🎉

ryan-son committed:
feat: Add select and deselect effects of cell

ryan-son committed:
feat: Add updating note feature

ryan-son committed:
refactor: Refactor structure of reflecting text changes

ryan-son committed:
feat: Connect to core data with `NoteCoreDataManager`

- Supports creating, reading, updating actions

ryan-son committed:
feat: Add trailing swipe action and remove action

- Alert presents when delete button tapped

ryan-son committed:
feat: Add leading swipe action and `activityViewTapped`

ryan-son committed:
fix: Fix cell layout bug when body only has new line string

ryan-son committed:
feat: Add actions when ellipsis is tapped

ryan-son committed:
refactor: Organize namespaces

ryan-son committed:
chore: Apply typealias to missed subjects

ryan-son committed:
chore: Remove unnecessary asset json and codes

ryan-son committed:
style: Change completion handler style to keep consistency

ryan-son committed:
chore: Change error message for loading persistent container

ryan-son committed:
chore: Subtitute codes with the existing method

ryan-son committed:
feat: Separate CRUD features from NoteListViewController to NoteManager

ryan-son committed:
chore: Add error cases and messages

ryan-son committed:
fix: Fix bugs when occurred while editing text

ryan-son committed:
feat: Develop CRUD features - completed

ryan-son committed:
chore: Modifiy error cases and messages

ryan-son committed:
chore: Remove unnecessary frameworks

ryan-son committed:
test: Add CRUD feature tests

ryan-son committed:
test: Add UI test cases

ryan-son committed:
refactor: Change delegation structures between types

ryan-son committed:
refactor: Organize codes for keyboard notifications

ryan-son committed:
feat: Change logic for adding and creating note

ryan-son committed:
chore: Add and apply error cases

ryan-son committed:
refactor: Refactor configuring delete button method

ryan-son committed:
test: Modify test methods with changed methods

ryan-son committed:
Merge pull request #8 from ryan-son/feature/step2/coredata-modelReplaced

Feature/step2/coredata

ryan-son committed:
fix: Resolve an issue for updating newly added note

ryan-son committed:
style: Change line wrapping for multiline string literal

ryan-son committed:
refactor: Separate text processing logic to String extension

ryan-son committed:
chore: Remove unnecessary delegate requirements

- Adjusted access levels of relevant properties and methods

ryan-son committed:
chore: Move `applySnapshot` method to main class

- `extension -> main` due to removed from protocol

ryan-son committed:
test: Change method for getting note from core data stack

ryan-son committed:
chore: Remove unnecessary method `getNote(byID:)`

ryan-son committed:
feat: Hide keyboard when more button is tapped

ryan-son commented:
## 딜리게이트 관련 추가 구조 변경에 대해 도미닉의 의견이 궁금합니다.
지금까지 delegate를 채택하여 구현하는 타입에 대한 오개념을 가지고 있었습니다. `UISplitViewControllerDelegate`, `UITextViewDelegate` 등 delegate를 채택하여 구현하는 타입은 해당 타입의 커스텀 타입 (예를 들어 `NoteTextView: UITextView`)이어야 한다고 생각했습니다. 하지만, `UITableViewDelegate`, `UITableViewDataSource`와 같이 `UIViewController` 타입이 delegate를 채택하여 사용했던 기억을 떠올려보니 잘못된 개념이었다는 것을 깨달았습니다.

그래서 `UITextViewDelegate`를  굳이 커스텀 타입인 `NoteTextView`가 구현하지 않고, 텍스트뷰를 `UITextView`로 생성한 후 `NoteDetailViewController`가 구현하여 자신이 `UITextView`의 딜리게이트가 되는 구조로 변경하면 어떨까요? 저는 구별하기 어려운 딜리게이트를 여러 개 생성하지 않고 구조를 단순하게 만들 수 있다는 측면에서 더 좋다고 생각합니다.
### 제안 구조
![image](https://user-images.githubusercontent.com/69730931/122201523-145f1c00-ced7-11eb-8725-ee888f8a2ae7.png)

### 종전 구조
![image](https://user-images.githubusercontent.com/69730931/122114478-484a2b00-ce5e-11eb-83d8-ee0dd5991532.png)

제안드린 구조의 코드
```swift
final class NoteDetailViewController: UIViewController {

    weak var noteListViewControllerDelegate: NoteListViewControllerDelegate?

    private var noteTextView: UITextView = {
        let noteTextView = UITextView()
        noteTextView.translatesAutoresizingMaskIntoConstraints = false
        noteTextView.font = UIFont.preferredFont(forTextStyle: .body)
        noteTextView.adjustsFontForContentSizeCategory = true
        noteTextView.accessibilityIdentifier = "noteTextView"
        return noteTextView
    }()

    ...

    override func viewDidLoad() {
        ...
        noteTextView.delegate = self
        ...
    }
}

extension NoteDetailViewController: UITextViewDelegate {
    func textViewDidChange(_ textView: UITextView) {
        noteListViewControllerDelegate?.applyTextUpdate(with: textView.text)
    }
}

protocol NoteListViewControllerDelegate: AnyObject {
    var currentIndexPathOfEditingNote: IndexPath { get }
    func deleteTapped(at indexPath: IndexPath)
    func activityViewTapped(at indexPath: IndexPath)
    func applyTextUpdate(with newText: String)
}
```

AppleCEO commented on CloudNotes/CloudNotes/Controllers/NoteDetailViewController.swift:
```diff
@@ -44,15 +66,19 @@ final class NoteDetailViewController: UIViewController {
         noteTextView.isEditable = false
+        moveTop(of: noteTextView)
+        setTextViewDelegatesIfNeeded()
     }
```
추상화 단계가 비슷한 코드들이 모여있어서 보기 편합니다! 💯

AppleCEO commented on CloudNotes/CloudNotes/Controllers/NoteDetailViewController.swift:
```diff
@@ -90,45 +117,121 @@ final class NoteDetailViewController: UIViewController {
-        textView.text = ViewContents.emptyString
-        textView.insertText(note.title + ViewContents.newLineString)
-        textView.insertText(note.body)
+        textView.text = note.title + UIItems.TextView.titleSeparatorString + note.body
```
이 부분도 매끄럽습니다~ 👍

AppleCEO commented on CloudNotes/CloudNotes/Controllers/NoteDetailViewController.swift:
```diff
@@ -90,45 +117,121 @@ final class NoteDetailViewController: UIViewController {
     
     private func moveTop(of textView: UITextView) {
-        textView.setContentOffset(CGPoint(x: 0, y: -view.safeAreaInsets.top), animated: false)
+        textView.setContentOffset(CGPoint(x: .zero, y: -view.safeAreaInsets.top), animated: false)
```
0과 .zero를 사용할 때 어떤 점이 다를까요?

ryan-son replied:
공식문서에도 zero는 덧셈 연산에서의 단순한 0의 값이라고 표현하고 있습니다. 따라서 앱 실행 간 동일한 동작을 보여줄 것입니다.

0과 같이 숫자로 표현하면 숫자 자체에는 의미가 없기 때문에 다른 프로그래머들이 코드를 작성한 사람의 의도를 이해하기 어려울 것이라 생각합니다. 왜 0을 의도한 것인지, 그 이외의 값은 안되는지에 대한 의문을 가질 수 있을 것이기 때문입니다.
하지만 `.zero`로 표현하면 읽는 사람의 입장에서도 확실히 값이 0임을 의도하였음을 알 수 있기 때문에 때문에 숫자로 작성하였을 때보다 0 이외의 값으로 섣불리 변경할 수 없는 느낌을 줄 수 있을 것이라 판단됩니다.

![image](https://user-images.githubusercontent.com/69730931/122365153-66667700-cf95-11eb-8f51-d56b0bdfbdd5.png)

AppleCEO commented on CloudNotes/CloudNotes/Controllers/NoteDetailViewController.swift:
```diff
@@ -90,45 +117,121 @@ final class NoteDetailViewController: UIViewController {
             setText(to: noteTextView, with: note)
             return
         }
-        
```
의미없는 공백을 제거한 부분 좋습니다 🥇

AppleCEO reviewed (commented):
안녕하세요. 도미닉입니다.
라이언 고생하셨습니다!

Unit test의 경우에는 in-memory 방식의 mock 타입을 작성하여 활용하는 방식으로 테스트 시 앱 본체의 core data와는 별도의 환경을 구성하여 사용하였는데, UI test 시에도 이러한 mock 타입 인스턴스를 사용하여 별도의 저장소에서 테스트할 수 있는 방법이 있을까요? UI 테스트로 인해 생성된 데이터가 실제 앱 실행 시에도 남아있어 문제가 있어 보입니다.
- UI Test는 안해봤지만 Unit Test 할 때 처럼 똑같이 주입해 줄 수 있지 않나요? UI Test 환경과 앱 실행환경이 구분이 안되는 건가요?

UI 테스트는 unit test와 달리 블랙박스 테스트라는 내용을 봤는데, 별도의 core data 저장소를 만들어 사용할 수 없다는 의미인가요?
코어 데이터를 이용한 CRUD 구현에 있어 현재 조직한 구조 (타입 간의 연관 관계)에 대해 어떻게 생각하시나요? 어떤 부분을 개선해보면 좋을까요?
- 코어 데이터 사용은 힘들다고 합니다. 관련된 테스트는 unit test로 진행하라고 하네요.[참고](https://stackoverflow.com/questions/36771561/how-can-i-access-a-core-data-store-from-an-xcode-ui-test)

텍스트를 수정하면 실시간으로 반영이 되게끔 구성하였는데요, 최초에는 Diffable data source의 스냅샷을 데이터 소스에 적용할 때 애니메이션 효과를 주고 싶었습니다. 하지만 셀의 순서 변경과 같은 애니메이션은 작동해도 데이터가 변경된 것이 셀에 반영되지는 않았는데, 어떤 실수를 하고 있는걸까요?
- 회의실에서 현상 공유해주셔서 파악에 용이했습니다! 같이 해결되서 다행이네요~ [링크](https://stackoverflow.com/questions/62995493/how-to-reload-data-in-section-in-uicollectionview-when-using-uicollectionviewdif)가 도움이 많이 됐네요! currentSnapshot.reloadSections([section])을 넣어줌으로서 해결이 돼서 다행입니다~

머지 원하시면 댓글이나 DM 주세요. 고생하셨습니다~

ryan-son committed:
feat: Add animating differences feature when updated

- Change delegate sturctures

ryan-son commented:
@AppleCEO 

도미닉, 피드백 감사드립니다.

snapshot의 section을 reload하여 수정사항이 발생할 때 즉시 셀에 반영되는 부분은 해결하였지만 어떠한 셀이든 삭제하면 런타임 에러가 발생하여 디버깅하는데 시간이 조금 걸렸습니다.
그래서 삭제되는 경우에는 reload 구문을 실행시키지 않도록 아래와 같이 코드를 조금 변경해보았습니다.

```swift
private func applySnapshot(animatingDifferences: Bool) {
    guard let dataSource = dataSource else {
        return
    }
    let sections = Section.allCases
    var updated = Snapshot()
    updated.appendSections(sections)
    updated.appendItems(noteManager.fetchedNotes)
    if let editingNote = editingNote, dataSource.snapshot().numberOfItems <= updated.numberOfItems {
        updated.reloadItems([editingNote])
    }
    dataSource.apply(updated, animatingDifferences: animatingDifferences)
}
```
`reloadSection`이 아닌 `reloadItems` 메서드를 사용한 이유는 변경된 아이템만 리로드하면 될 것이라 판단했기 때문입니다.

> UI Test는 안해봤지만 Unit Test 할 때 처럼 똑같이 주입해 줄 수 있지 않나요? UI Test 환경과 앱 실행환경이 구분이 안되는 건가요?

UI Test는 Unit test와 달리 `XCUIApplication()`을 통해 실제 앱을 대신(proxy)할 인스턴스를 생성하여 실행시킵니다. 하지만 사용하는 코어 데이터는 동일한지 테스트 후 남아있는 데이터가 있다면 테스트앱이 아닌 본체 앱에서도 데이터가 불러와집니다. 
Unit test는 테스트할 대상을 직접 인스턴스화하고 필요할 경우 특정 대상을 테스트용 객체로 변경할 수 있었지만, UI test에서는 그러한 조작은 할 수 없었고 앱을 실행하여 UI 동작에 대한 테스트만을 수행할 수 있었네요. 관련해서는 자세한 내용을 더 찾아보고 학습해보겠습니다!

AppleCEO commented:
고생하셨습니다~
